build.ps1

```diff
@@ -1,66 +1,140 @@
 
-$buildFolder = ".build"
-$buildFile="$buildFolder\KoreBuild.ps1"
+.PARAMETER DotNetHome
+The directory where .NET Core tools will be stored.
 
-if (!(Test-Path $buildFolder)) {
-    Write-Host "Downloading KoreBuild from $koreBuildZip"
+.PARAMETER ToolsSource
+The base url where build tools can be downloaded.
 
-    $tempFolder=$env:TEMP + "\KoreBuild-" + [guid]::NewGuid()
-    New-Item -Path "$tempFolder" -Type directory | Out-Null
+.PARAMETER Update
+Updates KoreBuild to the latest version even if a lock file is present.
 
-    $localZipFile="$tempFolder\korebuild.zip"
+.PARAMETER MSBuildArgs
+Arguments to be passed to MSBuild
 
-    DownloadWithRetry -url $koreBuildZip -downloadLocation $localZipFile -retries 6
+.NOTES
+This function will create a file $PSScriptRoot/korebuild-lock.txt. This lock file can be committed to source, but does not have to be.
+When the lockfile is not present, KoreBuild will create one using latest available version from $Channel.
 
-    Add-Type -AssemblyName System.IO.Compression.FileSystem
-    [System.IO.Compression.ZipFile]::ExtractToDirectory($localZipFile, $tempFolder)
+#>
+[CmdletBinding(PositionalBinding = $false)]
+param(
+    [Alias('p')]
+    [string]$Path = $PSScriptRoot,
+    [Alias('c')]
+    [string]$Channel = 'dev',
+    [Alias('d')]
+    [string]$DotNetHome = $(`
+            if ($env:DOTNET_HOME) { $env:DOTNET_HOME } `
+            elseif ($env:USERPROFILE) { Join-Path $env:USERPROFILE '.dotnet'} `
+            elseif ($env:HOME) {Join-Path $env:HOME '.dotnet'}`
+            else { Join-Path $PSScriptRoot '.dotnet'} ),
+    [Alias('s')]
+    [string]$ToolsSource = 'https://aspnetcore.blob.core.windows.net/buildtools',
+    [Alias('u')]
+    [switch]$Update,
+    [Parameter(ValueFromRemainingArguments = $true)]
+    [string[]]$MSBuildArgs
+)
 
-    New-Item -Path "$buildFolder" -Type directory | Out-Null
-    copy-item "$tempFolder\**\build\*" $buildFolder -Recurse
+Set-StrictMode -Version 2
+$ErrorActionPreference = 'Stop'
 
-    # Cleanup
-    if (Test-Path $tempFolder) {
-        Remove-Item -Recurse -Force $tempFolder
+#
+# Functions
+#
+
+function Get-KoreBuild {
+
+    $lockFile = Join-Path $Path 'korebuild-lock.txt'
+
+    if (!(Test-Path $lockFile) -or $Update) {
+        Get-RemoteFile "$ToolsSource/korebuild/channels/$Channel/latest.txt" $lockFile
     }
+
+    $version = Get-Content $lockFile -Tail 1
+    $korebuildPath = Join-Paths $DotNetHome ('buildtools', 'korebuild', $version)
+
+    if (!(Test-Path $korebuildPath)) {
+        Write-Host "Downloading KoreBuild $version"
+        New-Item -ItemType Directory -Path $korebuildPath | Out-Null
+        $remotePath = "$ToolsSource/korebuild/artifacts/$version/korebuild.$version.zip"
+
+        try {
+            if (Get-Command -Name 'Expand-Archive' -ErrorAction Ignore) {
+                # Use built-in commands where possible as they are cross-plat compatible
+                $tmpfile = "$(New-TemporaryFile).zip"
```

Comment:
[This calls `Path.GetTempFileName`](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs#L26) which creates the file it returns in advance. If `Expand-Archive` (and `ZipFile` below) don't care about the extension, we could just use it without creating extra files (though granted, they are 0-byte).

Alternatively, the way you create the temp file name in the other branch is totally safe, so there's no reason not to just do it here.

Wish we had `Path.GetTempFileName(string extension)`.